### PR TITLE
Implement issue #260 BSI Add() method and other enhancements

### PR DESCRIPTION
**Pull request by guymolinari**
New BSI Add() method added per issue #260.  Also, added Clone() and new transpose method.  Transpose methods are functional but not terribly performant.   Need to perhaps re-work these to use SIMD SSE2 or similar.

@lemire @jacksonrnewhouse - Please review ASAP.   Thanks.

Fixes https://github.com/RoaringBitmap/roaring/issues/260

### CODE REVIEW

**Comment by lemire** [comment]
The travis failure does not seem to be related to roaring (and this PR) but seems to have to do with something else.

File: BitSliceIndexing/bsi.go [hunk 1]
```diff
@@ -39,6 +40,23 @@ func NewDefaultBSI() *BSI {
 	return NewBSI(int64(0), int64(0))
 }
 
+func (b *BSI) RunOptimize() {
```
Comment (lemire) [inline BitSliceIndexing/bsi.go:43]: This needs to be documented if it is to be exported.

File: BitSliceIndexing/bsi.go [hunk 2]
```diff
@@ -39,6 +40,23 @@ func NewDefaultBSI() *BSI {
...
+	b.eBM.RunOptimize()
+	for i := 0; i < len(b.bA); i++ {
+		b.bA[i].RunOptimize()
+	}
+	b.runOptimized = true
+}
+
+func (b *BSI) HasRunCompression() bool {
```
Comment (lemire) [inline BitSliceIndexing/bsi.go:51]: This needs to be documented.

File: BitSliceIndexing/bsi.go [hunk 3]
```diff
@@ -39,6 +40,23 @@ func NewDefaultBSI() *BSI {
...
+	return b.runOptimized
+}
+
+// Returns a pointer to existence bitmap
+func (b *BSI) GetExistenceBitmap() *roaring.Bitmap {
+	return b.eBM
```
Comment (lemire) [inline BitSliceIndexing/bsi.go:57]: I think in Go, comments should be of the form "GetExistenceBitmap returns a ..."

File: BitSliceIndexing/bsi.go [hunk 4]
```diff
@@ -380,19 +472,35 @@ func (b *BSI) UnmarshalBinary(bitData [][]byte) error {
...
-			b.bA = append(b.bA, roaring.NewBitmap())
+			newBm := roaring.NewBitmap()
+			if b.runOptimized {
+				newBm.RunOptimize()
+			}
+			b.bA = append(b.bA, newBm)
 		}
 		if err := b.bA[i-1].UnmarshalBinary(bitData[i]); err != nil {
 			return err
+		} else {
```
Comment (lemire) [inline BitSliceIndexing/bsi.go:483]: This else is unnecessary. Rest of the code should not be in braces.

File: BitSliceIndexing/bsi.go [hunk 5]
```diff
@@ -497,3 +612,78 @@ func (b *BSI) NewBSIRetainSet(foundSet *roaring.Bitmap) *BSI {
 	wg.Wait()
 	return newBSI
 }
+
+func (b *BSI) Clone() *BSI {
```
Comment (lemire) [inline BitSliceIndexing/bsi.go:616]: If this is public, it should be documented.

File: BitSliceIndexing/bsi.go [hunk 6]
```diff
@@ -497,3 +612,78 @@ func (b *BSI) NewBSIRetainSet(foundSet *roaring.Bitmap) *BSI {
...
+	return b.NewBSIRetainSet(b.eBM)
+}
+
+// Add
+//
+// In-place sum the contents of another BSI with this BSI, column wise.
+//
+func (b *BSI) Add(other *BSI) {
```
Comment (lemire) [inline BitSliceIndexing/bsi.go:624]: The comment should be "Add does ..."

File: BitSliceIndexing/bsi.go [hunk 7]
```diff
@@ -497,3 +612,78 @@ func (b *BSI) NewBSIRetainSet(foundSet *roaring.Bitmap) *BSI {
...
+	}
+	resultsChan <- results
+}
+
+// Increment
+//
+// In-place increment of values in a BSI.  Found set select columns for incrementing.
+// It assumes that the column value already exists.
+//
+func (b *BSI) Increment(foundSet *roaring.Bitmap) {
```
Comment (lemire) [inline BitSliceIndexing/bsi.go:683]: The comment should be of the form "Increment does X"

File: roaring64/bsi64.go [hunk 8]
```diff
@@ -37,6 +39,23 @@ func NewDefaultBSI() *BSI {
 	return NewBSI(int64(0), int64(0))
 }
 
+func (b *BSI) RunOptimize() {
```
Comment (lemire) [inline roaring64/bsi64.go:42]: Undocumented exported function.

File: roaring64/bsi64.go [hunk 9]
```diff
@@ -37,6 +39,23 @@ func NewDefaultBSI() *BSI {
...
+	b.eBM.RunOptimize()
+	for i := 0; i < len(b.bA); i++ {
+		b.bA[i].RunOptimize()
+	}
+	b.runOptimized = true
+}
+
+func (b *BSI) HasRunCompression() bool {
```
Comment (lemire) [inline roaring64/bsi64.go:50]: Undocumented exported function.

File: roaring64/bsi64.go [hunk 10]
```diff
@@ -37,6 +39,23 @@ func NewDefaultBSI() *BSI {
...
+	return b.runOptimized
+}
+
+// Returns a pointer to existence bitmap
+func (b *BSI) GetExistenceBitmap() *Bitmap {
```
Comment (lemire) [inline roaring64/bsi64.go:55]: Comment does not have the expected format.

File: roaring64/bsi64.go [hunk 11]
```diff
@@ -378,19 +475,35 @@ func (b *BSI) UnmarshalBinary(bitData [][]byte) error {
...
-			b.bA = append(b.bA, NewBitmap())
+			newBm := NewBitmap()
+			if b.runOptimized {
+				newBm.RunOptimize()
+			}
+			b.bA = append(b.bA, newBm)
 		}
 		if err := b.bA[i-1].UnmarshalBinary(bitData[i]); err != nil {
 			return err
+		} else {
```
Comment (lemire) [inline roaring64/bsi64.go:486]: The else is unnecessary and the rest of the code should not be in braces.

File: BitSliceIndexing/bsi.go [hunk 12]
```diff
@@ -474,7 +585,11 @@ func (b *BSI) ClearValues(foundSet *roaring.Bitmap) {
 	wg.Wait()
 }
 
-// NewBSIRetainSet
+/*
+ * NewBSIRetainSet
+ *
+ * Construct a new BSI from a clone of existing BSI, retain only values contained in foundSet
+ */
 func (b *BSI) NewBSIRetainSet(foundSet *roaring.Bitmap) *BSI {
```
Comment (lemire) [inline BitSliceIndexing/bsi.go:593]: Comment on NewBSIRetainSet does not follow the expected format.

File: roaring64/bsi64.go [hunk 13]
```diff
@@ -495,3 +615,80 @@ func (b *BSI) NewBSIRetainSet(foundSet *Bitmap) *BSI {
 	wg.Wait()
 	return newBSI
 }
+
+func (b *BSI) Clone() *BSI {
```
Comment (lemire) [inline roaring64/bsi64.go:619]: Undocumented exported function.

File: roaring64/bsi64.go [hunk 14]
```diff
@@ -495,3 +615,80 @@ func (b *BSI) NewBSIRetainSet(foundSet *Bitmap) *BSI {
...
+	return b.NewBSIRetainSet(b.eBM)
+}
+
+// Add
+//
+// In-place sum the contents of another BSI with this BSI, column wise.
+//
+func (b *BSI) Add(other *BSI) {
```
Comment (lemire) [inline roaring64/bsi64.go:627]: Unexpected format in the comment.

File: roaring64/bsi64.go [hunk 15]
```diff
@@ -495,3 +615,80 @@ func (b *BSI) NewBSIRetainSet(foundSet *Bitmap) *BSI {
...
+	}
+	resultsChan <- results
+}
+
+// Increment
+//
+// In-place increment of values in a BSI.  Found set select columns for incrementing.
+// It assumes that the column value already exists.
+//
+func (b *BSI) Increment(foundSet *Bitmap) {
```
Comment (lemire) [inline roaring64/bsi64.go:688]: Unexpected format in the comment.

File: roaring64/bsi64.go [hunk 16]
```diff
@@ -495,3 +615,80 @@ func (b *BSI) NewBSIRetainSet(foundSet *Bitmap) *BSI {
...
+	b.addDigit(foundSet, 0)
+}
+
+func (b *BSI) IncrementAll() {
```
Comment (lemire) [inline roaring64/bsi64.go:692]: Undocumented exported function.

**Comment by happygiraffe** [comment]
The CI failure is because Travis is using Go 1.12, but the code uses `errors.Is()` which was [introduced](https://golang.org/doc/go1.13) in Go 1.13.

**Comment by lemire** [comment]
@happygiraffe Where does roaring use `errors.Is`?

**Comment by happygiraffe** [comment]
I was referring to this travis build:
- https://travis-ci.org/github/RoaringBitmap/roaring/builds/726112695

The error in the log is:

```
…
# github.com/stretchr/testify/assert
257../../stretchr/testify/assert/assertions.go:1703:5: undefined: errors.Is
258../../stretchr/testify/assert/assertions.go:1726:6: undefined: errors.Is
259../../stretchr/testify/assert/assertions.go:1749:5: undefined: errors.As
260../../stretchr/testify/assert/assertions.go:1766:7: undefined: errors.Unwrap
261../../stretchr/testify/assert/assertions.go:1770:7: undefined: errors.Unwrap
…
```

So, my bad: testify is using Go 1.13 features, not roaring. Sorry.

**Comment by lemire** [comment]
@happygiraffe Right. So testify breaks with older Go versions, and we use testify as part of some tests... but not in roaring per se.

File: BitSliceIndexing/bsi.go (hunk 2 above)
Comment (guymolinari) [reply BitSliceIndexing/bsi.go:51]: Done.

File: BitSliceIndexing/bsi.go (hunk 3 above)
Comment (guymolinari) [reply BitSliceIndexing/bsi.go:57]: Done

File: BitSliceIndexing/bsi.go (hunk 4 above)
Comment (guymolinari) [reply BitSliceIndexing/bsi.go:483]: Done

File: BitSliceIndexing/bsi.go (hunk 5 above)
Comment (guymolinari) [reply BitSliceIndexing/bsi.go:616]: Done.

File: BitSliceIndexing/bsi.go (hunk 7 above)
Comment (guymolinari) [reply BitSliceIndexing/bsi.go:683]: Done

File: roaring64/bsi64.go (hunk 8 above)
Comment (guymolinari) [reply roaring64/bsi64.go:43]: Done.

File: roaring64/bsi64.go (hunk 9 above)
Comment (guymolinari) [reply roaring64/bsi64.go:52]: Done.

File: roaring64/bsi64.go (hunk 10 above)
Comment (guymolinari) [reply roaring64/bsi64.go:57]: Done.

File: roaring64/bsi64.go (hunk 11 above)
Comment (guymolinari) [reply roaring64/bsi64.go]: Done.

File: BitSliceIndexing/bsi.go (hunk 12 above)
Comment (guymolinari) [reply BitSliceIndexing/bsi.go:593]: Fixed.

File: roaring64/bsi64.go (hunk 13 above)
Comment (guymolinari) [reply roaring64/bsi64.go:615]: Done.

File: roaring64/bsi64.go (hunk 14 above)
Comment (guymolinari) [reply roaring64/bsi64.go:620]: Fixed.

File: roaring64/bsi64.go (hunk 15 above)
Comment (guymolinari) [reply roaring64/bsi64.go:678]: Done.

File: roaring64/bsi64.go (hunk 16 above)
Comment (guymolinari) [reply roaring64/bsi64.go:683]: Done.

File: BitSliceIndexing/bsi.go (hunk 6 above)
Comment (lemire) [reply BitSliceIndexing/bsi.go:624]: This is still outstanding.

**Comment by lemire** [comment]
@guymolinari  We are still getting a bunch of warnings...

```
bsi.go:43:1: exported method BSI.RunOptimize should have comment or be unexported
bsi.go:51:1: exported method BSI.HasRunCompression should have comment or be unexported
bsi.go:55:1: comment on exported method BSI.GetExistenceBitmap should be of the form "GetExistenceBitmap ..."
bsi.go:483:10: if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)
bsi.go:500:9: if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)
bsi.go:588:1: comment on exported method BSI.NewBSIRetainSet should be of the form "NewBSIRetainSet ..."
bsi.go:616:1: exported method BSI.Clone should have comment or be unexported
bsi.go:620:1: comment on exported method BSI.Add should be of the form "Add ..."
bsi.go:678:1: comment on exported method BSI.Increment should be of the form "Increment ..."
bsi.go:687:1: exported method BSI.IncrementAll should have comment or be unexported
bsi64.go:619:1: comment on exported method BSI.Add should be of the form "Add ..."
bsi64.go:677:1: comment on exported method BSI.Increment should be of the form "Increment ..."
bsi64.go:682:1: comment on exported method BSI.IncrementAll should be of the form "IncrementAll ..."
```

May I am doing it wrong... but could you run `golint` on your side and see what it tells you? We are really trying to follow standard guidelines regarding code style.

Cheers!

**Comment by guymolinari** [comment]
@lemire Fixed all linting issues.

**Comment by lemire** [comment]
Merging.

**Pull request merged**